### PR TITLE
System replicated_fetches table

### PR DESCRIPTION
**Pull request by alesapin**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- New Feature

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Add `system.replicated_fetches` table which shows currently running background fetches.

### CODE REVIEW

File: tests/integration/test_system_replicated_fetches/test.py [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+    return ''.join(random.choice(string.ascii_uppercase + string.digits) for _ in range(length))
+
+def test_system_replicated_fetches(started_cluster):
+    node1.query("CREATE TABLE t (key UInt64, data String) ENGINE = ReplicatedMergeTree('/clickhouse/test/t', '1') ORDER BY tuple()")
+    node2.query("CREATE TABLE t (key UInt64, data String) ENGINE = ReplicatedMergeTree('/clickhouse/test/t', '2') ORDER BY tuple()")
+
+    with PartitionManager() as pm:
+        node2.query("SYSTEM STOP FETCHES t")
+        node1.query("INSERT INTO t SELECT number, '{}' FROM numbers(10000)".format(get_random_string(104857)))
+        pm.add_network_delay(node1, 80)
```
Comment (alesapin) [inline tests/integration/test_system_replicated_fetches/test.py:37]: Now we can add network delays in integration tests.

**Comment by alesapin** [comment]
```
2020.10.27 20:52:09.380775 [ 15249 ] {} <Debug> TCPHandler: Connected ClickHouse client version 20.11.0, revision: 54442, database: test_bdufo3, user: default.
2020.10.27 20:52:09.400145 [ 15249 ] {30df17a9-da35-4878-a6fa-2e942808424d} <Debug> executeQuery: (from [::1]:45012) DROP TABLE IF EXISTS alter_table_3
2020.10.27 20:52:09.400621 [ 15249 ] {30df17a9-da35-4878-a6fa-2e942808424d} <Trace> ContextAccess (default): Access granted: DROP TABLE ON test_bdufo3.alter_table_3
2020.10.27 20:52:09.400853 [ 15249 ] {30df17a9-da35-4878-a6fa-2e942808424d} <Trace> test_bdufo3.alter_table_3 (ReplicatedMergeTreeRestartingThread): Restarting thread finished
2020.10.27 20:52:09.403899 [ 15249 ] {30df17a9-da35-4878-a6fa-2e942808424d} <Trace> test_bdufo3.alter_table_3 (ReplicatedMergeTreeRestartingThread): Waiting for threads to finish
2020.10.27 20:52:09.403988 [ 15249 ] {30df17a9-da35-4878-a6fa-2e942808424d} <Information> test_bdufo3.alter_table_3 (cd086536-8ec4-46e5-ac47-ca543ad26e4e): Stopped being leader
2020.10.27 20:52:09.406985 [ 15249 ] {30df17a9-da35-4878-a6fa-2e942808424d} <Trace> test_bdufo3.alter_table_3 (ReplicatedMergeTreeRestartingThread): Threads finished
2020.10.27 21:01:01.288625 [ 11372 ] {0a853355-4403-4e18-b0e7-d0b1e9096100} <Debug> test_bdufo3.alter_table_3 (cd086536-8ec4-46e5-ac47-ca543ad26e4e): Waiting for r_7 to pull log-0000000220 to queue
2020.10.27 21:01:01.288855 [ 11372 ] {0a853355-4403-4e18-b0e7-d0b1e9096100} <Debug> test_bdufo3.alter_table_3 (cd086536-8ec4-46e5-ac47-ca543ad26e4e): Looking for node corresponding to log-0000000220 in r_7 queue
2020.10.27 21:01:01.315405 [ 161 ] {} <Trace> test_bdufo3.alter_table_5 (c25b14cd-7770-450e-a938-a9ff167fd20e): dropAllData: waiting for locks.
2020.10.27 21:01:01.315429 [ 11372 ] {0a853355-4403-4e18-b0e7-d0b1e9096100} <Debug> test_bdufo3.alter_table_3 (cd086536-8ec4-46e5-ac47-ca543ad26e4e): No corresponding node found. Assuming it has been already processed. Found 358 nodes
2020.10.27 21:01:01.315484 [ 161 ] {} <Trace> test_bdufo3.alter_table_5 (c25b14cd-7770-450e-a938-a9ff167fd20e): dropAllData: removing data from memory.
2020.10.27 21:01:01.315594 [ 11372 ] {0a853355-4403-4e18-b0e7-d0b1e9096100} <Debug> test_bdufo3.alter_table_3 (cd086536-8ec4-46e5-ac47-ca543ad26e4e): Finished waiting for all replicas to process log-0000000220
2020.10.27 21:01:01.687934 [ 174 ] {} <Information> DatabaseCatalog: Will try drop test_bdufo3.alter_table_3 (cd086536-8ec4-46e5-ac47-ca543ad26e4e)
2020.10.27 21:01:01.688114 [ 174 ] {} <Trace> test_bdufo3.alter_table_3 (ReplicatedMergeTreeRestartingThread): Restarting thread finished
2020.10.27 21:01:01.688219 [ 174 ] {} <Trace> test_bdufo3.alter_table_3 (ReplicatedMergeTreeRestartingThread): Waiting for threads to finish
2020.10.27 21:01:01.688288 [ 174 ] {} <Trace> test_bdufo3.alter_table_3 (ReplicatedMergeTreeRestartingThread): Threads finished
2020.10.27 21:01:01.688380 [ 174 ] {} <Information> test_bdufo3.alter_table_3 (cd086536-8ec4-46e5-ac47-ca543ad26e4e): Removing replica /clickhouse/tables/test_00993/alter_table/replicas/r_3, marking it as lost
2020.10.27 21:01:01.751207 [ 174 ] {} <Information> test_bdufo3.alter_table_3 (cd086536-8ec4-46e5-ac47-ca543ad26e4e): /clickhouse/tables/test_00993/alter_table/replicas/r_3 is the last replica, will remove table
2020.10.27 21:01:01.751471 [ 174 ] {} <Information> test_bdufo3.alter_table_3 (cd086536-8ec4-46e5-ac47-ca543ad26e4e): Removing table /clickhouse/tables/test_00993/alter_table (this might take several minutes)
2020.10.27 21:01:01.764014 [ 174 ] {} <Information> test_bdufo3.alter_table_3 (cd086536-8ec4-46e5-ac47-ca543ad26e4e): Table /clickhouse/tables/test_00993/alter_table was successfully removed from ZooKeeper
```
Seems like we wait for ALTER for too long after drop already called. I'll try to improve it, but this failure not related to changes.

File: src/Storages/MergeTree/DataPartsExchange.cpp [hunk 2]
```diff
@@ -52,6 +52,24 @@ std::string getEndpointId(const std::string & node_id)
...
+
+    explicit ReplicatedFetchReadCallback(ReplicatedFetchList::Entry & replicated_fetch_entry_)
+        : replicated_fetch_entry(replicated_fetch_entry_)
+    {}
+
+    void operator() (size_t bytes_count)
+    {
+        replicated_fetch_entry->bytes_read_compressed.store(bytes_count, std::memory_order_relaxed);
+        replicated_fetch_entry->progress.store(
+                static_cast<double>(bytes_count) / replicated_fetch_entry->total_size_bytes_compressed,
```
Comment (Avogar) [inline src/Storages/MergeTree/DataPartsExchange.cpp:68]: Is it possible that `total_size_bytes_compressed` is 0? I see that initial value is 0:
https://github.com/ClickHouse/ClickHouse/blob/1c9e24b5e2ab5c753b71a66e5d0d7c86e7df401c/src/Storages/MergeTree/DataPartsExchange.cpp#L285

Comment (alesapin) [reply src/Storages/MergeTree/DataPartsExchange.cpp:68]: Wow, yes. Of course, it's not possible, but initialization should be fixed.

**Review by Avogar — APPROVED** [review]

**Comment by filimonov** [comment]
Maybe we should have a unified table for background tasks? For now we can fill it with merges and fetches, later we can add moves, directory monitor, Kafka and others? If we will create one table per task type we may get too many of them...

**Comment by alesapin** [comment]
> Maybe we should have a unified table for background tasks? For now we can fill it with merges and fetches, later we can add moves, directory monitor, Kafka and others? If we will create one table per task type we may get too many of them...

Clear idea, it's more ClickHouse's way to store everything in a single table. But are there any other advantages? 
Disadvatages:
1) Such "background" tables may have a completely different structure.
2) It's much simplier to go to the system database and find table like `%fetches%` or `%kafka%` than try to guess which `WHERE` expression you have to add to filter required background processes.
3) If separate table exists, than we have information about corresponding background jobs. If we have a single table with all jobs then information may not exist or user just typed `WHERE` expression incorrectly.

**Comment by alesapin** [comment]
I'll merge it, if we will found some strong ideas against separate table then I'll change it before release.

**Comment by filimonov** [comment]
Related https://github.com/ClickHouse/ClickHouse/pull/7859

**Comment by filimonov** [comment]
From an operational perspective `select * from system.jobs` to understand what is happening right now sounds very cool.

There are alreade quite a lot of different types of background operations (dictionary reload, moves, fetches, sending parts, executing distributed ddl, rabbit, kafka, mutations, merges, 'special' types of merges - like TTL, checking parts, etc).  It would be cool to expose all of them one day.

> may have a completely different structure.

That is the strongest disadvantage. But Nested / json can be ok. 

> try to guess which WHERE expression ... If we have a single table with all jobs then information may not exist or user just typed WHERE expression incorrectly.

enum for operation type should prevent that problem 
```
SELECT * FROM system.query_log WHERE type = 'asdfasdf'

Received exception from server (version 20.10.3):
Code: 36. DB::Exception: Received from localhost:9000. DB::Exception: Unknown element 'asdfasdf' for type Enum8('QueryStart' = 1, 'QueryFinish' = 2, 'ExceptionBeforeStart' = 3, 'ExceptionWhileProcessing' = 4): while converting 'asdfasdf' to Enum8('QueryStart' = 1, 'QueryFinish' = 2, 'ExceptionBeforeStart' = 3, 'ExceptionWhileProcessing' = 4): while executing 'FUNCTION equals(type, 'asdfasdf') UInt8 = equals(type, 'asdfasdf')'. 
```

> If separate table exists, than we have information about corresponding background jobs. If we have a single table with all jobs then information may not exist or user just typed WHERE expression incorrectly.

Imagine you have some pick in a load for 0.5 sec every minute. You want to know what is that. Even if you will type something like 
```
SELECT * FROM system.merges
union all
SELECT * FROM system.moves
union all 
SELECT * FROM system.fetches 
union all 
...
```
you will catch the state of those tables in different moment of time. 

Actually maybe we can even have both? Overview-like system.jobs and per type of event (when more detailed info in needed)?

**Pull request merged**